src/Map/MapPrimitiviser_P.cpp

```diff
@@ -491,9 +491,24 @@ bool OsmAnd::MapPrimitiviser_P::determineSurfaceType(PointI center, QList<std::s
     // Rule: Water is always on the right along the direction of coastline segment.
     if (neareastCoastlineMapObject)
     {
-        const auto sign = crossProductSign(nearestCoastlineSegment0, nearestCoastlineSegment1, mCenter);
-        surfaceType = (sign >= 0) ? MapSurfaceType::FullLand : MapSurfaceType::FullWater;
-        return true;
+        // Filter too short nearest coastline for detailed maps.
+        // Rule: to correct calulation coastline should be greater then 1/10 of tilesize.
+        // Size of any detailed map tile is equal tile of ZoomLevel14.
```

Comment:
is it the same code in core-legacy? if not the bug should be reproducible in core-legacy